### ISSUE TITLE
Review 2

### ISSUE THREAD

**Issue by maxkfranz** [PathwayCommons/grounding-search#7]
- [ ] @maxkfranz Manually review the results of the grounding service
  - [ ] Are the results comparable to the raw Uniprot requests?
  - [x] Determine to what degree the aggregate service functionality in Factoid should be done manually versus by Elasticsearch in this project.
  - [x] Determine whether we should use a common index or one index per data source.  (I suspect a single index should be used, but let's see if I change my mind during review of other parts of the system.)
- [x] Things to revise
  - [x] @maxkfranz Add more basic logging to give feedback when updating the data.  It takes a long time, so it would be nice to know when
    - [x] downloading starts,
    - [x] downloading ends,
    - [x] processing starts,
    - [x] processing ends,
    - [x] indexing starts, and
    - [x] indexing ends.
  - [x] @maxkfranz If possible, it would be nice to have a scrollbar for each phase using [cli-progress](https://www.npmjs.com/package/cli-progress). -- Using plain log messages is enough now that things are faster.
  - [x] @metincansiper The code should be more general to use a single index.
    - [x] `search()` and `get()` will not need an index to differentiate datasources, but it will need a `namespace` filter.
    - [x] `clear()` and `update()` need to be aware of other data.  `clear()` should only remove the data for the particular namespace rather than the whole index.
    - [x] For `uniprot.update()`, we should probably be inserting to Elasticsearch each entry, or entries in small batches, so that we don't maintain a large `entries` array.
    - [x] XML parsing is very slow.  We should try a different lib (for example, [saxes](https://www.npmjs.com/package/saxes))
  - [ ] @maxkfranz We should use the `_score` result for the organism tie-breaking logic.  Because the result size is relatively small (20-50, say), it's probably fine to do this outside of Elasticsearch.  I don't see a straightforward way to incorporate our organism tie-breaking into an Elasticsearch query.

**Comment by metincansiper** [comment]
> XML parsing is very slow. We should try a different lib (for example, saxes)

I tested performance of different xml readers. Three of them looks worth mentioning here.

- [saxes](https://www.npmjs.com/package/saxes): Fastest of the libraries that I tried so far (read the data in 325s 133.757205ms in my first run). However, how it handles data is not good for our usage, maybe this is one of the reasons why it is so fast. ``onopentag`` gets a node parameter that only has "name" and "attributes" fields. We can still try to read the value of node by using ``ontext`` but it does not omit the white spaces and requires extra effort for that. Also, we are not able to keep track of parent child relationships with this library. I think that using this library will require unsecure workarounds even if it is hardly possible.

- [xml-streamer](https://www.npmjs.com/package/xml-streamer): One of the fastest libraries though it is outperformed by saxes (processed data in 804s 311.878834ms in my first run, 20s faster in the second run if I am not wrong). This library looks like a good fit for our usage too. Objects properties are mapped by keys with multi level of nesting.

- [xml-reader](https://www.npmjs.com/package/xml-reader): Has slightly better performance than xml-streamer (691s 206.601156ms in first run). However, does not return an object as well defined as in xml-streamer case. All children are piled up in ``children`` array. So it will require extra effort to handle it and most probably this extra effort will already be spending the slightly better performance than xml-streamer.

Considering that this operation will be performed rarely on server side I would favor reliability over a little better performance. Therefore, I think that xml-streamer is the best fit for our usage. @maxkfranz what do you think?

**Comment by maxkfranz** [comment]
@metincansiper Performance is very important, because we may need to use very large datasources.  (As an example, Pubchem is about 250GB of compressed data, 400GB of compressed data if you use both substances+compounds for mapping.)  It needs to be as fast as possible, even if the code is more complicated.

The docs for saxes isn't that good, but the events are pretty much the same as for sax.

The important ones are:

- `opentag`
- `text`
- `attribute`
- `closetag`

Basically, you just maintain a stack.  You push when you get a new `opentag` and you pop when you get a `closetag`.   You also make a ref in the parent to the new child on `opentag`.  I can help you get started with this once we decide on a lib.

[htmlparser2](https://www.npmjs.com/package/htmlparser2) is another option.

I would rather use the fastest of the most widely used (streaming/evented) XML parsing libs and build the JSON ourselves.  It's important that it's well tested and well maintained.

We have the option of offloading the JSON processing to another thread, so the raw parsing performance of each lib is very important.  Even a difference of 2x can make a big difference if the dataset is big.

**Comment by maxkfranz** [comment]
Here is a minimal test case for uniprot data.  It includes only the subset of the data that we care about.

```
<entry dataset="Swiss-Prot" created="2006-03-21" modified="2019-02-13" version="161">
  <accession>Q7LG56</accession>
  <accession>B4E2N4</accession>
  <accession>Q17R22</accession>
  <accession>Q75PQ6</accession>
  <accession>Q75PQ7</accession>
  <accession>Q75PY8</accession>
  <accession>Q75PY9</accession>
  <accession>Q86YE3</accession>
  <accession>Q9NPD6</accession>
  <accession>Q9NTD8</accession>
  <accession>Q9NUW3</accession>
  <name>RIR2B_HUMAN</name>
  <protein>
    <recommendedName>
      <fullName>Ribonucleoside-diphosphate reductase subunit M2 B</fullName>
      <ecNumber>1.17.4.1</ecNumber>
    </recommendedName>
    <alternativeName>
      <fullName>TP53-inducible ribonucleotide reductase M2 B</fullName>
    </alternativeName>
    <alternativeName>
      <fullName>p53-inducible ribonucleotide reductase small subunit 2-like protein</fullName>
      <shortName>p53R2</shortName>
    </alternativeName>
  </protein>
  <gene>
    <name type="primary">RRM2B</name>
    <name type="synonym">P53R2</name>
  </gene>
  <organism>
    <name type="scientific">Homo sapiens</name>
    <name type="common">Human</name>
    <dbReference type="NCBI Taxonomy" id="9606" />
  </organism>
</entry>
```

There are only a few tags we care about under `entry`:

- `accession`
- `name`
- `protein`
- `gene`
- `organism`

The rest can be ignored for building JSON.

**Comment by metincansiper** [comment]
Oh ``onclosetag`` was missing in their documentation I tried after reading your comment and yes they actually provide it. I think as long as we are able to listen to open and close tag events it is possible to figure out the nested structure of object precisely, which is fine.

**Comment by metincansiper** [comment]
Just to make a correction ``onclosetag`` was not actually missing in the documentation. However, it is not directly written in doc page so I missed it.

**Comment by metincansiper** [comment]
@maxkfranz saxes looks faster then htmlparser2 as well (the same file takes around 166ms for saxes and around 221ms for htmlparser2). I tried numerious other libs too saxes was the fastest one. Can we decide on saxes? If so I can start implementation for that library.

**Comment by metincansiper** [comment]
I think what to do is clear in my mind no need for a starter point, I can ask questions if I stuck at some point. I was pessimistic about ``saxes`` because I could not find ``closetag`` event initially. It looks fine as we have that event.

**Comment by metincansiper** [comment]
We will be needing the same logic for other data sources as well. Therefore, looks like it is useful to write a component at the top of xml stream library (looks like most probably ``saxes``). The main function of the component would be creating a json object based on the given input and triggering an event (or alternatively taking a callback function and consuming it) when close tag of the root path (in this case uniprot/entry) is encountered and even function would be able to consume the created json object. Example input maybe as the fallowing:

```
[
   {
      rootPath: 'uniprot/entry',
      collect: [
          { path: 'organism/dbReference', omitSelf: true, attributes: [ 'id' ] },
          { path: 'accession', singleValue: true },
          { path: 'gene/name' },
          ...
      ]
   },
  ...
]
```

``omitSelf: true`` means that the self text (e.g. ``Q7LG56`` for  ``<accession>Q7LG56</accession>``) of that path must not be collected. For example for organism/dbReference we are only interested in id attribute.
``singleValue: true`` means that only the first occurrence of path will be considered and its value will be collected as single value, not inside an array. For example, for accession we are only interested in the first occurrence.

An object can be formed by collecting these paths (the ones that we are interested in) until the close tag of root path is encountered and the resulting object can be consumed by the callback function (either through triggering an event or passing the callback function as option). In the future other data sources can also utilize the same component as long as their input format is xml.

**Comment by maxkfranz** [comment]
I wouldn't worry about trying to make it general yet.  Just focus on making it work with easy to read code.

**Comment by metincansiper** [comment]
I made changes for speeding up XML parsing process. As we talked I used ``saxes`` library as the parser. There is a significant improvement in speed. I counted execution time of uniprot update in 2 runs (excluding downloading file). One of them took 403 s 924.109846 ms while the other one took 380 s 954.677679 ms. Keep in mind that when I counted time of only parsing the same file using ``saxes`` (no processing at all just running the parser)  it took 325s 133.757205ms (I did it a while ago).

**Comment by metincansiper** [comment]
Looks like even with the same parameters the run time may vary in matter of 10 secs so it is hard to compare. But I can say for the run time calculations that I stated in above comment the general performance was at the worse of it and the pure xml library performance was at the best of it. 

On average I can say pure lib parsing (with no further processing cost) takes around 350s-360s and whole executions takes around 380s (in my laptop). The difference may even more or less on average. Not able to make precise observations.

Also, looks like updating chunk size did not make any visible performance improvement. I tried to see if I can use larger static arrays to ensure that there will not be any dynamic size duplication while using bigger chunks, so less db IO. However, neither that improved performance (I suppose maybe db operations was already not taking considerable amount of time).
